ingest/template/schema_parser.py

```diff
@@ -0,0 +1,266 @@
+import json
+import re
+
+import jsonref
+
+from ingest.utils import doctict
+from .exceptions import RootSchemaException
+
+
+class SchemaParser:
+    """ A SchemaParser provides functions for accessing objects in a JSON schema. """
+
+    def __init__(self, template):
+        self.ignored_properties = ["describedBy", "schema_version", "schema_type", "provenance"]
+        self.required_properties = []
+
+        self.schema_template = template
+
+        # TODO: identifiable should be in the schema - hardcoded here for now.
+        self._identifiable = ["biomaterial_id", "process_id", "protocol_id", "file_name"]
+
+        self._key_lookup = {}
+
+    def load_schema(self, json_schema):
+        """
+        Resolve references in a given JSON-formatted metadata schema and populate a SchemaTemplate object with the
+        data in the metadata schema.
+
+        :param json_schema: An object representing a deserialized JSON-formatted metadata schema with references.
+        :return: An object representing a deserialized JSON-formatted metadata schema with all its references resolved.
+        """
+
+        # Use jsonrefs to resolve all $refs in JSON
+        metadata_schema_data = jsonref.loads(json.dumps(json_schema))
+        return self.__initialise_template(metadata_schema_data)
+
+    def load_migration(self, property_migration):
+        return self.__initialise_property_migration_template(property_migration)
+
+    def key_lookup(self, key):
+        return self._key_lookup[key]
+
+    def __initialise_property_migration_template(self, property_migration):
+        migrated_property = property_migration["source_schema"] + "." + property_migration["property"]
+
+        migration_info = {}
+
+        if "target_schema" in property_migration and "replaced_by" in property_migration:
+            migration_info["replaced_by"] = \
+                property_migration["target_schema"] + "." + property_migration["replaced_by"]
+
+        if "effective_from" in property_migration:
+            migration_info["version"] = property_migration["effective_from"]
+        elif "effective_from_source" in property_migration:
+            migration_info["version"] = property_migration["effective_from_source"]
+            migration_info["target_version"] = property_migration["effective_from_target"]
+
+        migration_info = {migrated_property.split(".")[-1]: migration_info}
+        for part in reversed(migrated_property.split(".")[:-1]):
+            migration_info = {part: migration_info}
+
+        self.schema_template.put_migration(migration_info)
+        return self.schema_template
+
+    def __initialise_template(self, data):
```

Comment:
Can these method names be refactored to not start with two underscores? The appropriate naming convention is to start with one underscore; two underscores are reserved for language-internal or binding-internal methods.

Reply:
Removed all the underscores from now until I can figure out what exactly needs to be private/public.